src/vm/prestub.cpp

```diff
@@ -250,90 +243,346 @@ void DACNotifyCompilationFinished(MethodDesc *methodDesc)
 #endif
 // </TODO>
 
+PCODE MethodDesc::PrepareInitialCode()
+{
+    STANDARD_VM_CONTRACT;
+    return PrepareCode(GetInitialCodeVersion());
+}
 
-// ********************************************************************
-//                  README!!
-// ********************************************************************
+PCODE MethodDesc::PrepareCode(NativeCodeVersion codeVersion)
+{
+    STANDARD_VM_CONTRACT;
+    PrepareCodeConfig config;
+    if (codeVersion.IsDefaultVersion())
+    {
+        config = PrepareCodeConfig(codeVersion);
+    }
+#if FEATURE_CODE_VERSIONING
+    else
+    {
+        CodeVersionManager::TableLockHolder lock(GetCodeVersionManager());
+        config = PrepareCodeConfig(codeVersion);
+    }
+#endif
+    return PrepareCode(&config);
+}
 
-// MakeJitWorker is the thread safe way to invoke the JIT compiler
-// If multiple threads get in here for the same pMD, ALL of them
-// MUST return the SAME value for pstub.
-//
-// This function creates a DeadlockAware list of methods being jitted
-// which prevents us from trying to JIT the same method more that once.
+PCODE MethodDesc::PrepareCode(PrepareCodeConfig* pConfig)
+{
+    STANDARD_VM_CONTRACT;
+    pConfig->FinishConfiguration();
 
+    // If other kinds of code need multi-versioning we could add more cases here,
+    // but for now generation of all other code/stubs occurs in other code paths
+    _ASSERTE(IsIL() || IsNoMetadata());
+    return PrepareILBasedCode(pConfig);
+}
 
-PCODE MethodDesc::MakeJitWorker(COR_ILMETHOD_DECODER* ILHeader, CORJIT_FLAGS flags)
+PCODE MethodDesc::PrepareILBasedCode(PrepareCodeConfig* pConfig)
 {
     STANDARD_VM_CONTRACT;
+    PCODE pCode = NULL;
 
-    BOOL fIsILStub = IsILStub();        // @TODO: understand the need for this special case
+    if (pConfig->MayUsePrecompiledCode())
+    {
+        pCode = GetPrecompiledCode(pConfig);
+    }
+    if (pCode == NULL)
+    {
+        LOG((LF_CLASSLOADER, LL_INFO1000000,
+            "    In PrepareILBasedCode, calling JitCompileCode\n"));
+        // Mark the code as hot in case the method ends up in the native image
+        g_IBCLogger.LogMethodCodeAccess(this);
```

Comment:
Is it intentional to only call this when pCode == NULL?  I take it that we're not worried about logging the access here unconditionally because in precompiled code, the probes will get executed?